libraries/api/src/middleware/auth.js

```diff
@@ -246,6 +233,7 @@ const remoteAuthentication = ({ signatureValidator, entry }) => {
         // occurs with getArtifact for a public artifact, for example)
         if (!scopeExpression) {
           req.public = true;
+          req.hasAuthed = true;
```

Comment:
have they really authemticated in this case? I feel like this should still be false unless I'm misremembering the meaning of this flag.

Reply:
This is a good question. The docs defined `hasAuthed` as _Will be true if the request had `req.authorize()` called on it._ In this case, `req.authorize` has been called, but there again, those docs were written when the `no-auth` case was handled outside of `req.authorize` so this could probably be cleaned up.

The places where `noAuth`'s value is tested are:

1) in `libraries/api/src/middleware/handle.js` where an error is thrown if the API call is not public, and `hasAuthed` is `false`
2) in `libraries/api/src/middleware/logging.js` to only retrieve `clientId` and `expires` if `hasAuthed` is true
3) in `libraries/api/src/middleware/schema.js` where an error is thrown if the API call is not public, and `hasAuthed` is `false` (this seems to be a duplication of point 1)

So pre-RFC-165, there was a simple relationship between whether a method was public, and whether it was necessary for the client's authorization credentials to be authorized: namely, if a method wasn't public (i.e. the rendered scope template resulted in a scopeExpression containing one of more scopes), then the client request had to contain hawk authentication, and be authorized against the required scopes. However, post-RFC-165, it is possible for a non-authorized client request to call a guarded API, so long as the required scopes for the API are satisfied by the scopes of the anonymous role.

So if we were to consider `hasAuthed` to mean whether the full set of scopes available to the API request have been determined, then it would still fulfill its function for 1) and 3) above. For 2) I think it is sufficient to use a `clientId` if one is returned, and the same for `expires`, and not depend on the `hasAuthed` property. We can possibly remove the check in 1 or 3 if we decide they are equivalent (although I would be a little anxious to do this directly in this PR where there is already a lot of other auth change).

Reply:
I've fixed this up now in later commits.